### DISCUSSION TITLE
Trouble with instancing Recursive's rvrn monospace character substitutions

### DISCUSSION THREAD

**Discussion by Shrinks99** [fonttools/fonttools#3057] [Q&A] [answered]
Hi there!

This started out as an issue in [Slice](https://github.com/source-foundry/Slice/issues/109) that I'll ask here because they seem to use fontTools on the back end. :)

I'm trying to create a static version of [Recursive](https://www.recursive.design/) set to specific values to be used as a brand asset.  Recursive has a mono axis which swaps out certian glyphs with a `.mono` rvrn single substituion lookup (#248 in the font if you're gonna go looking for it).  For most apps that I use that support variable fonts (Figma, Windows' font viewer, Inkscape) the instanced file created by fontforge with the `varLib.instancer` command will correctly result in the expected output.  In tools like Affinity Designer that don't support variable fonts (and therefore don't support the rvrn OpenType function) this results in an instanced file where the non .mono suffix glyphs are used as if the file didn't have its mono axis set to 0.51+.

For reproducing this, the command I used was: `fonttools varLib.instancer ./Recursive_VF_1.085.ttf MONO=0.52 CASL=0 slnt=0 CRSV=0 wght=715`

## Examples of Output Issues

Here's an example of what this looks like in Figma.  The top text is stock variable Recursive, the bottom is an instanced version of the font.  In Figma everything looks proper!

![227051731-97055d21-6475-4324-a8c6-645c5693f992](https://user-images.githubusercontent.com/5672810/227428903-a0fbb929-3224-4882-a01c-15e0cce1fa88.png)

In Affinity Designer however using the same font file, the .mono glyphs are not being used, note the different r & l

![227051803-94fb6278-90ab-4d3b-9cea-b2217b6b4f03](https://user-images.githubusercontent.com/5672810/227429020-5b663f02-9c55-4706-a6a7-0a968b731c02.png)

The monospaced r & l shown above can't be accessed in Affinity Designer through a stylistic set or any other solution because it doesn't support the variable features Recurisve uses to swap the glyphs, even as a static cut.  Is there a way to get fontTools to instance the file with the mono characters used for everything by default?  Is there utility to including rvrn rules in a static cut of a typeface or should I file this as a bug report?  Any help would be much appreciated :)

**Comment by jenskutilek** [comment, marked as answer]
Hm, the [spec](https://learn.microsoft.com/en-us/typography/opentype/spec/features_pt#tag-rvrn) says:

> Application of the 'rvrn' feature is mandatory in implementations that support OpenType Font Variations whenever a variable font is in use. The feature should be processed in any layout process that supports use of variations, even if other OpenType Layout processing is not supported.

After instancing, the font is not a variable font anymore, so it seems Affinity is not at fault here when they ignore the `rvrn` feature. `varLib.instancer` perhaps should drop the feature completely and change the affected glyph mapping the the `cmap` table.

As a workaround, you could change the glyph mapping after instancing by “freezing” the `rvrn` feature, e.g. with [pyfeatfreeze](https://github.com/twardoch/fonttools-opentype-feature-freezer).

**Reply by Shrinks99** [reply]
This was _exactly_ the tool I needed to solve my issue!  Thank you so much :)

As for my opinion, after going through this experience I would say that `varLib.instancer` would ideally take the `rvrn` rules into account when generating a single cut of a font?  Would have saved me some trouble, that's for sure!

**Comment by anthrotype** [comment]
yes, the instancer simply copies the Feature record of the FeatureVariations that apply at the given location to the default FeatureList so that it will be always activated, with the assumption that a shaping engine will apply 'rvrn' when it seesn one; we don't change the FeatureTag. That could also be an option maybe, e.g. trying to rename "rvrn" to "ccmp" which is similarly applied early in shaping, and ought to be supported even when variations are not.

> “freezing” the rvrn feature pyfeatfreeze

yes, if that works, I suggest you do that.

**Reply by behdad** [reply]
> yes, if that works, I suggest you do that.

I think we should do that when instantiating.

**Reply by behdad** [reply]
> I think we should do that when instantiating.

Or move its lookups to `ccmp`. That would be easier.
